lib/db/jdb-internal.c

```diff
@@ -0,0 +1,324 @@
+j_db_delete_exec(JList* operations, JSemantics* semantics)
+{
+	return j_backend_db_func_exec(operations, semantics, J_MESSAGE_DB_DELETE);
+}
+gboolean
+j_db_internal_delete(gchar const* namespace, gchar const* name, bson_t const* selector, JBatch* batch, GError** error)
+{
+	JOperation* op;
+	JBackend_db_operation_data* data;
+	data = g_slice_new(JBackend_db_operation_data);
+	memcpy(data, &j_db_delete_params, sizeof(JBackend_db_operation_data));
+	data->in_param[0].ptr_const = namespace;
+	data->in_param[1].ptr_const = name;
+	data->in_param[2].ptr_const = selector;
+	data->out_param[0].ptr_const = error;
+	op = j_operation_new();
+	op->key = namespace;
+	op->data = data;
+	op->exec_func = j_db_delete_exec;
+	op->free_func = j_backend_db_func_free;
+	j_batch_add(batch, op);
+	return TRUE;
+}
+static gboolean
+j_db_get_all_exec(JList* operations, JSemantics* semantics)
+{
+	return j_backend_db_func_exec(operations, semantics, J_MESSAGE_DB_GET_ALL);
+}
+gboolean
+j_db_internal_query(gchar const* namespace, gchar const* name, bson_t const* selector, gpointer* iterator, JBatch* batch, GError** error)
+{
+	J_db_iterator_helper* helper;
+	JOperation* op;
+	JBackend_db_operation_data* data;
+	j_goto_error_backend(!iterator, JULEA_BACKEND_ERROR_ITERATOR_NULL, "");
+	helper = g_slice_new(J_db_iterator_helper);
+	helper->initialized = FALSE;
+	memset(&helper->bson, 0, sizeof(bson_t));
+	*iterator = helper;
+	data = g_slice_new(JBackend_db_operation_data);
+	memcpy(data, &j_db_get_all_params, sizeof(JBackend_db_operation_data));
+	data->in_param[0].ptr_const = namespace;
+	data->in_param[1].ptr_const = name;
+	data->in_param[2].ptr_const = selector;
+	data->out_param[0].ptr_const = &helper->bson;
+	data->out_param[1].ptr_const = error;
+	op = j_operation_new();
+	op->key = namespace;
+	op->data = data;
+	op->exec_func = j_db_get_all_exec;
+	op->free_func = j_backend_db_func_free;
+	j_batch_add(batch, op);
+	return TRUE;
+_error:
+	return FALSE;
+}
+gboolean
+j_db_internal_iterate(gpointer iterator, bson_t* metadata, GError** error)
+{
+	gint ret;
+	const uint8_t* data;
+	uint32_t length;
+	J_db_iterator_helper* helper = iterator;
+	bson_t zerobson;
+	memset(&zerobson, 0, sizeof(bson_t));
+	if (!helper->initialized)
+	{
+		if (!memcmp(&helper->bson, &zerobson, sizeof(bson_t)))
+		{
+			g_set_error(error, JULEA_BACKEND_ERROR, JULEA_BACKEND_ERROR_BSON_INVALID, "");
+			goto error2;
+		}
+		bson_iter_init(&helper->iter, &helper->bson);
+		helper->initialized = TRUE;
+	}
+	ret = bson_iter_next(&helper->iter);
+	j_goto_error_backend(!ret, JULEA_BACKEND_ERROR_ITERATOR_NO_MORE_ELEMENTS, "");
+	ret = BSON_ITER_HOLDS_DOCUMENT(&helper->iter);
+	j_goto_error_backend(!ret, JULEA_BACKEND_ERROR_BSON_INVALID_TYPE, bson_iter_type(&helper->iter));
+	bson_iter_document(&helper->iter, &length, &data);
+	bson_init_static(metadata, data, length);
+	return TRUE;
+_error:
+	bson_destroy(&helper->bson);
+error2:
+	g_slice_free(J_db_iterator_helper, helper);
+	return FALSE;
+}
```

Comment:
Is there a reason for this to be in a separate source file?

Reply:
In this file are client-side functions which are direct wrappers for server-side functions.
Declaring these functions separately allows "direct" testing of the backend without the client side library on top.